docs: added out-of-order and out-of-bound receiver troubleshoot cases;

ethernoy opened the pull request:
<!--
    Keep PR title verbose enough and add prefix telling
    about what components it touches e.g "query:" or ".*:"
-->

<!--
    Don't forget about CHANGELOG!

    Changelog entry format:
    - [#<PR-id>](<PR-URL>) Thanos <Component> ...

    <PR-id> Id of your pull request.
    <PR-URL> URL of your PR such as https://github.com/thanos-io/thanos/pull/<PR-id>
    <Component> Component affected by your changes such as Query, Store, Receive.
-->

* [X] I added CHANGELOG entry for this change.
* [X] Change is not relevant to the end user.

## Changes

<!-- Enumerate changes you made -->
- Added Receiver section in troubleshoot document.
- Added troubleshoot case for out-of-bound Receiver error.
- Added troubleshoot case for out-of-order Receiver error.

## Verification

<!-- How you tested it? How do you know it works? -->
- Tested the solution in my own cluster

wiardvanrij commented on docs/operating/troubleshooting.md:
```diff
@@ -78,3 +78,74 @@ global:
+
+- Remote Prometheus is running in high availability mode (more than 1 replica are running). But the replica_external_label_name is not correctly configured (e.g. empty).
+
+![](../img/thanos_receiver_troubleshoot_empty_replica_external_label_name.png)
```
Please add alt text to these images with a description of the image. For example:

```
![Example view of compactor not coping with amount and size of incoming blocks](compactor_no_coping_with_load.png)
``` 
This improves accessibility :)

wiardvanrij commented:
This is really awesome! Love it. 
I checked the preview (https://deploy-preview-4134--thanos-io.netlify.app/tip/operating/troubleshooting.md/#out-of-bound-error) and I think the images are a bit out of bounds. 

I actually have no idea how to fix this other than scaling the images ourself to a certain max-width. (I will ask this on Slack if anyone got an idea) 
Other than that, LGTM!

ethernoy commented:
> This is really awesome! Love it.
> I checked the preview (https://deploy-preview-4134--thanos-io.netlify.app/tip/operating/troubleshooting.md/#out-of-bound-error) and I think the images are a bit out of bounds.
> 
> I actually have no idea how to fix this other than scaling the images ourself to a certain max-width. (I will ask this on Slack if anyone got an idea)
> Other than that, LGTM!

Thanks! I will resolve the accessibility issue tonight (I'm in UTC+8). Please let me know if you have any suggestions on how to resize the image from markdown point of view, if it can't be resolved here, I will consider resizing the image file. By the way, do you think it is a good idea to put the source .drawio file of those diagrams in the repository?

ethernoy commented:
> This is really awesome! Love it.
> I checked the preview (https://deploy-preview-4134--thanos-io.netlify.app/tip/operating/troubleshooting.md/#out-of-bound-error) and I think the images are a bit out of bounds.
> 
> I actually have no idea how to fix this other than scaling the images ourself to a certain max-width. (I will ask this on Slack if anyone got an idea)
> Other than that, LGTM!

Hi, I updated the PR:

- added several alt text to images for better accessibility
- added fluid-img css class to images to improve the style of the page (now it won't go off-border)

wiardvanrij reviewed (dismissed):
LGTM, really neat.

ethernoy commented on CHANGELOG.md:
```diff
@@ -15,6 +15,7 @@ We use _breaking :warning:_ to mark changes that are not backward compatible (re
 ### Added
 
 - [#4107](https://github.com/thanos-io/thanos/pull/4107) Store: `LabelNames` and `LabelValues` now support label matchers.
+- [#4134](https://github.com/thanos-io/thanos/pull/4134) Docs: add out-of-order and out-of-bound receiver troubleshoot cases.
```
sure, will remove it in next commit

ethernoy commented on docs/operating/troubleshooting.md:
```diff
@@ -78,3 +78,74 @@ global:
+### Possible Cause
+
+1. Thanos Receiver was stopped previously and is just resumed, remote Prometheus starts to write from the oldest sample, which is too old to be digested and hence rejected.
+1. Thanos Receiver does not have enough resources to ingest the remote write data. The latest ingested sample is gradually falling behind the latest scraped samples.
```
> Can we also add a case from your amazing diagrams? (: (Multiple receivers adding sample for the same series without unique external labels (e.g replica labels).

I can add a diagram, but need more information for this case. Do you mean there will be `Error on ingesting samples that are too old or are too far into the future` when:


- there is one Prometheus remote writing to multiple receivers
- those receivers does not configure external_label
- those receivers does not configure a hashring (thus request for the same serie are handled by different receivers)

bwplotka replied:
I think I missed the error, I meant the other error case, so all good!

bwplotka reviewed (dismissed):
Solid and super helpful. Thank you!

bwplotka commented on docs/operating/troubleshooting.md:
```diff
@@ -78,3 +78,74 @@ global:
+
+- Remote Prometheus is running in high availability mode (more than 1 replica are running). But the replica_external_label_name is not correctly configured (e.g. empty).
+
+<img src="../img/thanos_receiver_troubleshoot_empty_replica_external_label_name.png" class="img-fluid" alt="Example topology diagram of out-of-order error case caused by empty replica_external_label_name" />
```
BTW, how you created those diagrams? If Google diagrams, do you mind linking the Google drawing and sharing edit rights with Thanos maintainers for easier editions? (:

bwplotka replied:
you can use our email `thanos-io@googlegroups.com`

ethernoy replied:
> BTW, how you created those diagrams? If Google diagrams, do you mind linking the Google drawing and sharing edit rights with Thanos maintainers for easier editions? (:

I used draw.io, the source files are in xml format. I don't mind sharing those with you guys, maybe I can just sent those source files to the email group in attachment?

wiardvanrij replied:
Maybe add drawio as SVG? WIth https://marketplace.visualstudio.com/items?itemName=hediet.vscode-drawio we can then just IN EDITOR edit svg's. Which contains BOTH the drawio base AND the image itself. 
Pretty..neat with some overhead of-course

ethernoy replied:
> Maybe add drawio as SVG? WIth https://marketplace.visualstudio.com/items?itemName=hediet.vscode-drawio we can then just IN EDITOR edit svg's. Which contains BOTH the drawio base AND the image itself.
> Pretty..neat with some overhead of-course

I can save those drawIO files as svg/png (with drawIO source inside, which is its feature) in the thanos repository and refer to them directly in the troubleshoot document, will do it ASAP.

ethernoy replied:
added

bwplotka commented:
Thank you!
